### DISCUSSION TITLE
rocketmq broker会占用多个端口？

### DISCUSSION THREAD

**Discussion by Hoxxx** [apache/rocketmq#6208] [Q&A] [answered]
在broker配置了listenPort=30911，启动后发现30909和30912都被rocketmq占用了，这是正常的吗？

**Comment by Hoxxx** [comment]
现在还有个新问题，broker group 2改了端口后（从30911改成了31911），mqadmin ClusterList 看不到broker group 2，想知道rocketMQ controller支持下面这种架构吗？
![image](https://user-images.githubusercontent.com/2639432/221792956-7dce187f-0e1f-482e-8cd9-2c9c088cbce2.png)

**Reply by Hoxxx** [reply]
这个问题估计是namesrv里metadata的问题，把namesrv的controllerStorePath全部删了重建貌似就好了
第一次提示地址被使用；第二次改了端口显示不出来；第三次把controllerStorePath全部干掉重建就显示出来了

**Comment by Hoxxx** [comment]
另外想问下mqshutdown broker能指定broker吗？

**Comment by ShadowySpirits** [comment, marked as answer]
> 在broker配置了listenPort=30911，启动后发现30909和30912都被rocketmq占用了，这是正常的吗？

30909 被占用是因为启动了 fastRemotingServer

https://github.com/apache/rocketmq/blob/88000dac212a0c721a99ca167132920ce0f45f1b/broker/src/main/java/org/apache/rocketmq/broker/BrokerController.java#L429-L435

30912 被占用是供 HA 使用

https://github.com/apache/rocketmq/blob/88000dac212a0c721a99ca167132920ce0f45f1b/container/src/main/java/org/apache/rocketmq/container/BrokerContainerStartup.java#L140

> 另外想问下mqshutdown broker能指定broker吗？

不能，mqshutdown 是通过 `ps` 来获取 broker pid 的

https://github.com/apache/rocketmq/blob/88000dac212a0c721a99ca167132920ce0f45f1b/distribution/bin/mqshutdown#L19-L30
